### PR TITLE
Add clipping overload to a iso-cuboid

### LINKED ISSUE #4415: Intersecting a mesh with a box
# Issue Details

This is not a bug, but rather a feature request. It would be really great if CGAL would provide a simple function to intersect a mesh with a box. It looks like a basic kind of functionality.

I am aware of the undocumented internal function clip_to_bbox(), yet it is not clear why this function also takes a plane as an input, in addition to the desired box.

I am also aware of the CGAL::Polygon_mesh_processing::clip() function that intersects two meshes. I tried to make a cube mesh and intersect it with my mesh, but that resulted in some weird mesh that meshlab refused to load complaining about some eof thing.

I know that such geometric computation is a hard thing, but hey, you folks are experts. :) Ideally a function like this should be provided, and it should automatically clean up any degeneracy, such as when the cutting plane intersects a face within numerical precision when bad things can happen like tiny edges, degenerate triangles, etc, etc. 

I am so pleased with CGAL so far, and I hope it keeps on improving.

### PR DESCRIPTION
**Pull request by sloriot**
fix #4415

### CODE REVIEW

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/clip.h [hunk 1]
```diff
@@ -432,6 +433,70 @@ bool clip(      TriangleMesh& tm,
                                       np, CGAL::graph_has_property<TriangleMesh, CGAL::face_index_t>());
 }
 
+/**
+  * \ingroup PMP_corefinement_grp
+  * clips `tm` by keeping the part that is inside `iso_cuboid`.
+  * If `tm` is closed, the clipped part can be closed too if the named parameter `clip_volume` is set to `true`.
+  * See subsection \ref coref_clip for more details.
```
Comment (afabri) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/clip.h]: Subsection

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/clip.h [hunk 2]
```diff
@@ -432,6 +433,70 @@ bool clip(      TriangleMesh& tm,
...
+  *
+  * \note In the current implementation it is not possible to set the vertex point map and the default will be used. `Iso_cuboid_3` must be
+  * from the same Kernel as the point of the vertex point map.
```
Comment (afabri) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/clip.h]: Kernel will get a link

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/clip.h [hunk 3]
```diff
@@ -432,6 +433,70 @@ bool clip(      TriangleMesh& tm,
...
+  *                            that is used to track the creation of new faces.
+  *   \cgalParamEnd
+  *   \cgalParamBegin{throw_on_self_intersection} if `true`,
+  *      the set of triangles closed to the intersection of `tm` and `plane` will be
+  *      checked for self-intersections and `CGAL::Polygon_mesh_processing::Corefinement::Self_intersection_exception`
+  *      will be thrown if at least one is found.
+  *   \cgalParamEnd
+  *   \cgalParamBegin{clip_volume} if `true` and `tm` is closed, the clipping will be done on
+  *      the volume \link coref_def_subsec bounded \endlink by `tm` rather than on its surface
+  *      (i.e. `tm` will be kept closed).
```
Comment (afabri) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/clip.h]: put a comma after i.e.

**Comment by oleg-alexandrov** [comment]
Thank you for your attention.

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/clip.h [hunk 4]
```diff
@@ -432,6 +433,70 @@ bool clip(      TriangleMesh& tm,
...
+  *   \cgalParamBegin{clip_volume} if `true` and `tm` is closed, the clipping will be done on
+  *      the volume \link coref_def_subsec bounded \endlink by `tm` rather than on its surface
+  *      (i.e., `tm` will be kept closed).
+  *   \cgalParamEnd
+  *   \cgalParamBegin{use_compact_clipper} if `false` and `clip_volume` is `false` and `tm` is open, the parts of `tm` coplanar with `plane`
+  *                                        will not be part of the output.
+  * \cgalNamedParamsEnd
+  *
+  * @return `true` if the output surface mesh is manifold.
+  *         If `false` is returned `tm` is only refined by the intersection with `plane`.
```
Comment (MaelRL) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/clip.h]: looks wrong

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/clip.h [hunk 5]
```diff
@@ -432,6 +433,70 @@ bool clip(      TriangleMesh& tm,
...
+  *   \cgalParamBegin{throw_on_self_intersection} if `true`,
+  *      the set of triangles closed to the intersection of `tm` and `plane` will be
+  *      checked for self-intersections and `CGAL::Polygon_mesh_processing::Corefinement::Self_intersection_exception`
+  *      will be thrown if at least one is found.
+  *   \cgalParamEnd
+  *   \cgalParamBegin{clip_volume} if `true` and `tm` is closed, the clipping will be done on
+  *      the volume \link coref_def_subsec bounded \endlink by `tm` rather than on its surface
+  *      (i.e., `tm` will be kept closed).
+  *   \cgalParamEnd
+  *   \cgalParamBegin{use_compact_clipper} if `false` and `clip_volume` is `false` and `tm` is open, the parts of `tm` coplanar with `plane`
```
Comment (sloriot) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/clip.h]:
```suggestion
  *   \cgalParamBegin{use_compact_clipper} if `false` and `clip_volume` is `false` and `tm` is open, the parts of `tm` coplanar with `is_cuboid`
```

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/clip.h [hunk 6]
```diff
@@ -432,6 +433,70 @@ bool clip(      TriangleMesh& tm,
...
+  *   \cgalParamBegin{clip_volume} if `true` and `tm` is closed, the clipping will be done on
+  *      the volume \link coref_def_subsec bounded \endlink by `tm` rather than on its surface
+  *      (i.e., `tm` will be kept closed).
+  *   \cgalParamEnd
+  *   \cgalParamBegin{use_compact_clipper} if `false` and `clip_volume` is `false` and `tm` is open, the parts of `tm` coplanar with `is_cuboid`
+  *                                        will not be part of the output.
+  * \cgalNamedParamsEnd
+  *
+  * @return `true` if the output surface mesh is manifold.
+  *         If `false` is returned `tm` is only refined by the intersection with `plane`.
```
Comment (sloriot) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/clip.h]:
```suggestion
  *         If `false` is returned `tm` is only refined by the intersection with `iso_cuboid`.
```

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/clip.h [hunk 7]
```diff
@@ -432,6 +433,70 @@ bool clip(      TriangleMesh& tm,
...
+  * @param tm input triangulated surface mesh
+  * @param iso_cuboid iso-cuboid used to clip `tm`.
+  * @param np optional sequence of \ref pmp_namedparameters "Named Parameters" among the ones listed below
+  *
+  * \cgalNamedParamsBegin
+  *   \cgalParamBegin{visitor} a class model of `PMPCorefinementVisitor`
+  *                            that is used to track the creation of new faces.
+  *   \cgalParamEnd
+  *   \cgalParamBegin{throw_on_self_intersection} if `true`,
+  *      the set of triangles closed to the intersection of `tm` and `plane` will be
```
Comment (sloriot) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/clip.h]:
```suggestion
  *      the set of triangles closed to the intersection of `tm` and `iso_cuboid` will be
```

**Comment by lrineau** [comment]
Tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-70.shtml and Ic-68

**Pull request merged**